sdk/core/System.ClientModel/src/ModelReaderWriter/SequenceBufferWriter.cs

```diff
@@ -0,0 +1,141 @@
+    /// </summary>
+    /// <param name="sizeHint">The minimum length of the returned <see cref="Memory{T}"/>. If less than 256, a buffer of size 256 will be returned.</param>
+    /// <returns>A memory buffer of at least <paramref name="sizeHint"/> bytes. If <paramref name="sizeHint"/> is less than 256, a buffer of size 256 will be returned.</returns>
+    public Memory<byte> GetMemory(int sizeHint = 0)
+    {
+        if (sizeHint < 256)
+        {
+            sizeHint = 256;
+        }
+
+        int sizeToRent = sizeHint > _segmentSize ? sizeHint : _segmentSize;
+
+        if (_buffers.Length == 0)
+        {
+            ExpandBuffers(sizeToRent);
+        }
+
+        ref SequenceBuffer last = ref _buffers[_count - 1];
+        Memory<byte> free = last.Array.AsMemory(last.Written);
+        if (free.Length >= sizeHint)
+        {
+            return free;
+        }
+
+        // else allocate a new buffer:
+        ExpandBuffers(sizeToRent);
+
+        return _buffers[_count - 1].Array;
+    }
+
+    private readonly object _lock = new object();
+    private void ExpandBuffers(int sizeToRent)
+    {
+        lock (_lock)
+        {
+            int bufferCount = _count == 0 ? 1 : _count * 2;
+
+            SequenceBuffer[] resized = new SequenceBuffer[bufferCount];
+            if (_count > 0)
+            {
+                _buffers.CopyTo(resized, 0);
+            }
+            _buffers = resized;
+            _buffers[_count].Array = ArrayPool<byte>.Shared.Rent(sizeToRent);
+            _count = bufferCount == 1 ? bufferCount : _count + 1;
+        }
+    }
+
+    /// <summary>
+    /// Returns a <see cref="Span{T}"/> to write to that is at least the requested size, as specified by the <paramref name="sizeHint"/> parameter.
+    /// </summary>
+    /// <param name="sizeHint">The minimum length of the returned <see cref="Span{T}"/>. If less than 256, a buffer of size 256 will be returned.</param>
+    /// <returns>A buffer of at least <paramref name="sizeHint"/> bytes. If <paramref name="sizeHint"/> is less than 256, a buffer of size 256 will be returned.</returns>
+    public Span<byte> GetSpan(int sizeHint = 0)
+    {
+        Memory<byte> memory = GetMemory(sizeHint);
+        return memory.Span;
+    }
+
+    /// <summary>
+    /// Disposes the SequenceWriter and returns the underlying buffers to the pool.
+    /// </summary>
+    public void Dispose()
+    {
+        int bufferCountToFree;
+        SequenceBuffer[] buffersToFree;
+        lock (_lock)
+        {
+            bufferCountToFree = _count;
+            buffersToFree = _buffers;
+            _count = 0;
+            _buffers = Array.Empty<SequenceBuffer>();
+        }
+
+        for (int i = 0; i < bufferCountToFree; i++)
+        {
+            ArrayPool<byte>.Shared.Return(buffersToFree[i].Array);
+        }
+    }
+
+    public SequenceBufferReader GetSequenceBufferReader()
```

Comment:
Could we call it ExtractReadOnlySequence to indicate a bit better what's happening.